source/simulate/profile-simulation.md

```diff
@@ -0,0 +1,647 @@
+def parse_fragment(fragment: str) -> int:
+    return int(fragment)
+```
+
+Let's see how it does:
+
+```
+Line #  Hits   Time        Per Hit    % Time         Line Contents
+==================================================================
+
+355       50   1809954.0   36199.1     99.0          parameters_at_instant = self.parameters.get_at_instant(str(instant))
+16     44100    421208.0       9.6     49.0          instant = str(periods.instant(instant))
+41     44100     97584.0       2.2     20.6          fragments = [parse_fragment(fragment) for fragment in fragments]
+212       17        40.0       2.4    100.0          return int(fragment)
+```
+
+Wow! The number of hits to `parse_fragment` decreased by 99%, improving the generator performance by 353%!
+
+That's it!
+
+## Caching further
+
+Once we know it works, can't we go further?
+
+With a couple of extra caches and code refactoring, we could end up with something like this:
+
+```python
+from __future__ import annotations
+
+# ...
+
+import functools
+
+# ...
+
+from typing import Optional
+
+# ...
+
+@functools.lru_cache(maxsize = 1000)
+def instant(instant) -> Optional[periods.Instant]:
+
+    if instant is None:
+        return None
+
+    if isinstance(instant, str):
+        if not config.INSTANT_PATTERN.match(instant):
+            raise_error(instant)
+
+        fragments = instant.split('-', 2)[:3]
+        integers = [parse_fragment(fragment) for fragment in fragments]
+        instant = periods.Instant(integers)
+
+    elif isinstance(instant, datetime.date):
+        instant = periods.Instant((instant.year, instant.month, instant.day))
+
+    elif isinstance(instant, int):
+        instant = (instant,)
+
+    elif isinstance(instant, (list, tuple)):
+        assert 1 <= len(instant) <= 3
+        instant = tuple(instant)
+
+    elif isinstance(instant, periods.Period):
+        instant = instant.start
+
+    elif isinstance(instant, periods.Instant):
+        return instant
+
+    else:
+        return instant
+
+    return globals()[f"instant_{len(instant)}"](instant)
+
+
+@functools.lru_cache(maxsize = 1000)
+def parse_fragment(fragment: str) -> int:
+    return int(fragment)
+
+
+@functools.lru_cache(maxsize = 1000)
+def instant_1(instant: periods.Instant) -> periods.Instant:
+    return periods.Instant((instant[0], 1, 1))
+
+
+@functools.lru_cache(maxsize = 1000)
+def instant_2(instant: periods.Instant) -> periods.Instant:
+    return periods.Instant((instant[0], instant[1], 1))
+
+
+@functools.lru_cache(maxsize = 1000)
+def instant_3(instant: periods.Instant) -> periods.Instant:
+    return periods.Instant(instant)
+```
+
+Please note that we're importing `annotations` from `__future__` so as to postpone the evaluation of type annotations ([PEP484](https://www.python.org/dev/peps/pep-0484/) and [PEP563](https://www.python.org/dev/peps/pep-0563/)) so as to avoid [cyclic imports](https://discuss.python.org/t/how-to-add-type-hints-for/7459).
```

Comment:
```suggestion
Please note that we're importing `annotations` from `__future__` so as to postpone the evaluation of type annotations ([PEP484](https://www.python.org/dev/peps/pep-0484/) and [PEP563](https://www.python.org/dev/peps/pep-0563/)), in order to avoid [cyclic 

> imports](https://discuss.python.org/t/how-to-add-type-hints-for/7459).

```